Odata-docs/odatalib/whats-new/odatalib-8.md

```diff
@@ -0,0 +1,429 @@
+
+Using `$filter` when you intend to fetch an entity by key can lead to surprising results. For example, consider the following snippet of code:
+
+```csharp
+var context = new Container(new Uri("https://services.odata.org/TripPinRESTierService/"));
+var friend = context.People.Where(p => p.UserName == "russellwhyte").Select(p => p.BestFriend);
+
+Console.WriteLine(friend.FirstName);
+```
+
+In OData .NET 7, the following OData URL would be generated by default: `https://services.odata.org/TripPinRESTierService/People('russellwhyte')/BestFriend`. Notice that the selected `BestFriend` is added as a navigation property segment to the URL. This request returns a the entity corresponding to the best friend at the top level of the response:
+
+```json
+{
+    "@odata.context": "https://services.odata.org/TripPinRESTierService/(S(q5f0myjzjtfondm5l1a5qsbw))/$metadata#People/$entity",
+    "UserName": "scottketchum",
+    "FirstName": "Scott",
+    "LastName": "Ketchum",
+    ...
+}
+```
+
+However, in OData .NET 8, this code would generate the following URL: `https://services.odata.org/TripPinRESTierService/People?$filter=UserName eq 'russellwhyte'&$expand=BestFriend`. Notice that request filters the entity set collection and also includes the navigation property using the `$expand` query. This means the result will be a collection containing a single entity, which has an expanded `BestFriend` property:
+
+```csharp
+{
+    "@odata.context": "https://services.odata.org/TripPinRESTierService/(S(pbicrumqvat4igher2mbadb3))/$metadata#People(BestFriend())",
+    "value": [
+        {
+            "UserName": "russellwhyte",
+            "FirstName": "Russell",
+            "LastName": "Whyte",
+            ...
+            "BestFriend": {
+                "UserName": "scottketchum",
+                "FirstName": "Scott",
+                "LastName": "Ketchum",
+                ...
+            }
+        }
+    ]
+}
+```
+
+This means that trying to access the result with `friend.FirstName` will fail and throw an exception, because the result is not the friend entity, but a collection that contains the filtered entity with the `BestFriend` property attached to it. Therefore in OData .NET 8 we would have to adjust the code as follows:
+
+```csharp
+// fetch people with expanded navigation property using OData URI People?$expand=BestFriend
+var people = await context.People.Expand('BestFriend').ExecuteAsync();
+// perform a local LINQ transform on the result data
+var bestFriends = people.Select(p => p.BestFriend);
+
+Console.WriteLine(bestFriends.First().FirstName);
+```
+
+Alternatively we could use the `ByKey` method:
+
+```csharp
+var friend = await ctx.People.ByKey("russelwhyte").Select(p => p.BestFriend).GetValueAsync();
+Console.WriteLine(friend.FirstName);
+```
+
+Alternatively, we could set `DataServiceContext.KeyComparisonGeneratesFilterQuery` to false, but it's no longer recommended:
```

Comment:
I think we can get rid of this statement since we do not recommend this so that people get used to the other options.

Reply:
The reason I mentioned it here is for people for whom this change will break things, and migration might require changing a lot of code (if the old pattern was used in many places). This could provide a quick _temporary_ workaround.